client_plugin/drivers/shared/etcdops.go

```diff
@@ -72,60 +77,75 @@ type kvPair struct {
 	value string
 }
 
-// initEtcd start or join ETCD cluster depending on the role of the node
-func (d *VolumeDriver) initEtcd() error {
+type etcdKVS struct {
+	driver   *VolumeDriver
+	nodeID   string
+	nodeAddr string
+	client   *etcdClient.Client
+}
+
+// NewKvStore function: start or join ETCD cluster depending on the role of the node
+func NewKvStore(driver *VolumeDriver) *etcdKVS {
+	var e *etcdKVS
+
 	ctx := context.Background()
-	docker := d.dockerd
+	docker := driver.dockerd
 
 	// get NodeID from docker client
 	info, err := docker.Info(ctx)
 	if err != nil {
 		log.WithFields(
 			log.Fields{"error": err},
 		).Error("Failed to get Info from docker client ")
-		return err
+		return nil
 	}
 
-	// worker just returns
+	// get the swarmID and IP address of current node
 	nodeID := info.Swarm.NodeID
+	addr := info.Swarm.NodeAddr
+
+	e = &etcdKVS{
+		driver:   driver,
+		nodeID:   nodeID,
+		nodeAddr: addr,
+	}
+
+	// worker just returns
 	if info.Swarm.ControlAvailable == false {
 		log.WithFields(
 			log.Fields{"nodeID": nodeID},
 		).Info("Swarm node role: worker. Action: return from InitEtcd ")
-		return nil
+		return e
 	}
 
 	// check my local role
 	node, _, err := docker.NodeInspectWithRaw(ctx, nodeID)
 	if err != nil {
 		log.WithFields(log.Fields{"nodeID": nodeID,
 			"error": err}).Error("Failed to inspect node ")
-		return err
+		return nil
 	}
 
-	// get the IP address of current node
-	addr := info.Swarm.NodeAddr
-
 	// if leader, proceed to start ETCD cluster
 	if node.ManagerStatus.Leader {
 		log.WithFields(
 			log.Fields{"nodeID": nodeID},
 		).Info("Swarm node role: leader, start etcd cluster")
-		err = startEtcdCluster(addr, nodeID)
+		err = e.startEtcdCluster()
 		if err != nil {
 			log.WithFields(log.Fields{"nodeID": nodeID,
 				"error": err}).Error("Failed to start ETCD Cluster")
-			return err
+			return nil
 		}
-		return nil
+		return e
 	}
 
 	// if manager, first find out who's leader, then proceed to join ETCD cluster
 	nodes, err := docker.NodeList(ctx, types.NodeListOptions{})
 	if err != nil {
 		log.WithFields(log.Fields{"nodeID": nodeID,
 			"error": err}).Error("Failed to get NodeList from swarm manager")
-		return err
+		return nil
```

Comment:
For the first question:
It's OK because only swarm **leader** will call the startEtcdCluster, and non-leader managers won't call startEtcdCluster.

For the second question:
No. Because swarm leader doesn't need to call NodeList. The workflow is: if this node is a leader, startEtcdCluster and leave; if this node is a non-leader manager, do NodeList to find who's the leader, then joinEtcdCluster and leave.